core/src/main/java/org/apache/druid/java/util/http/client/pool/ResourcePool.java

```diff
@@ -189,17 +190,63 @@ private ImmediateCreationResourceHolder(
         );
       }
     }
+  }
+
+  private static class LazyCreationResourceHolder<K, V> extends ResourceHolderPerKey<K, V>
+  {
+    private LazyCreationResourceHolder(
+        int maxSize,
+        long unusedResourceTimeoutMillis,
+        K key,
+        ResourceFactory<K, V> factory
+    )
+    {
+      super(maxSize, unusedResourceTimeoutMillis, key, factory);
+    }
+  }
+
+  private static class ResourceHolderPerKey<K, V> implements Closeable
+  {
+    protected final int maxSize;
+    private final K key;
+    private final ResourceFactory<K, V> factory;
+    private final long unusedResourceTimeoutMillis;
+    // Hold previously created / returned resources
+    protected final ArrayDeque<ResourceHolder<V>> resourceHolderList;
+    // Maintains count of times when get() threw an exception.
```

Comment:
can you add more details @AmatyaAvadhanula. Is it just a counter? What is its purpose